src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs

```diff
@@ -1716,36 +1775,98 @@ private Conversion ClassifyImplicitNullableConversion(TypeSymbol source, TypeSym
             return Conversion.NoConversion;
         }
 
-        private Conversion ClassifyImplicitTupleConversion(TypeSymbol source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        private delegate Conversion ClassifyConversionFromExpressionDelegate(BoundExpression sourceExpression, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics, bool arg);
+        private delegate Conversion ClassifyConversionFromTypeDelegate(TypeSymbol source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics, bool arg);
+
+        private Conversion GetImplicitTupleLiteralConversion(BoundTupleLiteral source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
         {
-            ImmutableArray<TypeSymbol> sourceTypes;
-            ImmutableArray<TypeSymbol> destTypes;
+            return GetTupleLiteralConversion(
+                source,
+                destination,
+                ref useSiteDiagnostics,
+                ConversionKind.ImplicitTupleLiteral,
+                (BoundExpression s, TypeSymbol d, ref HashSet<DiagnosticInfo> u, bool a) => ClassifyImplicitConversionFromExpression(s, d, ref u),
+                arg: false);
+        }
 
-            if (!source.TryGetElementTypesIfTupleOrCompatible(out sourceTypes) ||
-                !destination.TryGetElementTypesIfTupleOrCompatible(out destTypes) ||
-                sourceTypes.Length != destTypes.Length)
+        private Conversion GetExplicitTupleLiteralConversion(BoundTupleLiteral source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics, bool forCast)
+        {
+            return GetTupleLiteralConversion(
+                source,
+                destination,
+                ref useSiteDiagnostics,
+                ConversionKind.ExplicitTupleLiteral,
+                (BoundExpression s, TypeSymbol d, ref HashSet<DiagnosticInfo> u, bool a) => ClassifyConversionFromExpression(s, d, ref u, a),
+                forCast);
+        }
+
+        private static Conversion GetTupleLiteralConversion(
+            BoundTupleLiteral source,
+            TypeSymbol destination,
+            ref HashSet<DiagnosticInfo> useSiteDiagnostics,
+            ConversionKind kind,
+            ClassifyConversionFromExpressionDelegate classifyConversion,
```

Comment:
Would it be simpler to pass options, rather than a delegate? Seems like there are three cases: implicit, explicit and implicit-for-this-in-extension. Maybe two bools, or a small enum?